File: Python/Problem_122.py
def solve(MAX_K: int = 200) -> int:
    INF = 10**9
    best = [INF] * (MAX_K + 1)
    best[1] = 0

    # quick check: are we done?
    def done():
        for k in range(1, MAX_K + 1):
            if best[k] == INF:
                return False
        return True

    # depth-limited DFS building increasing addition chains
    def dfs(chain, depth, limit):
        x = chain[-1]
        if depth < best[x]:
            best[x] = depth

        if depth == limit:
            return

        # remaining steps
        rem = limit - depth

        # generate next candidates: x + chain[i] (i from end to start)
        # descending i gives larger nxt earlier (often helps pruning)
        for i in range(len(chain) - 1, -1, -1):
            nxt = x + chain[i]
            if nxt <= x or nxt > MAX_K:
                continue

            nd = depth + 1
            if nd >= best[nxt]:
                continue

            # bound: with rem-1 steps left, max reachable is nxt * 2^(rem-1)
            # if that can't reach MAX_K, still might help smaller ks, so we do NOT prune by MAX_K reach.
            # (If you only cared about one target, you'd prune harder.)

            chain.append(nxt)
            dfs(chain, nd, limit)
            chain.pop()

    # empirical upper bound: m(200) is small (< 12), but just loop until finished
    limit = 0
    while True:
        dfs([1], 0, limit)
        if done():
            return sum(best[1:])  # include m(1)=0
        limit += 1

if __name__ == "__main__":
    print(solve(200))  # 1582

File: CPP/Problem_122.cpp
#include <bits/stdc++.h>
using namespace std;

static const int MAX_K = 200;
static const int INF = 1e9;

int bestArr[MAX_K + 1];

static bool done() {
    for (int k = 1; k <= MAX_K; k++) {
        if (bestArr[k] == INF) return false;
    }
    return true;
}

static void dfs(vector<int>& chain, int depth, int limit) {
    int x = chain.back();
    if (depth < bestArr[x]) bestArr[x] = depth;
    if (depth == limit) return;

    for (int i = (int)chain.size() - 1; i >= 0; i--) {
        int nxt = x + chain[i];
        if (nxt <= x || nxt > MAX_K) continue;
        int nd = depth + 1;
        if (nd >= bestArr[nxt]) continue;

        chain.push_back(nxt);
        dfs(chain, nd, limit);
        chain.pop_back();
    }
}

int main() {
    for (int i = 0; i <= MAX_K; i++) bestArr[i] = INF;
    bestArr[1] = 0;

    int limit = 0;
    while (true) {
        vector<int> chain;
        chain.push_back(1);
        dfs(chain, 0, limit);
        if (done()) break;
        limit++;
    }

    long long sum = 0;
    for (int k = 1; k <= MAX_K; k++) sum += bestArr[k];
    cout << sum << "\n"; // 1582
    return 0;
}
